File: JavaScript/5_DataStructures/4_Queues.js
class SimpleQueue {
  constructor() {
    this.queue = [];
  }
  enqueue(val) {
    this.queue.push(val);
  }
  dequeue() {
    this.queue.shift();
  }
}

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }

  printAll() {
    let queue = [];
    if (!this.first) return null;
    let node = this.first;
    while (node) {
      queue.push(node.val);
      node = node.next;
    }
    return queue;
  }

  enqueue(val) {
    const new_node = new Node(val);
    if (!this.first) {
      this.first = new_node;
      this.last = new_node;
    } else {
      this.last.next = new_node;
      this.last = new_node;
    }
    return ++this.size;
  }

  dequeue() {
    if (!this.first) return -1;
    let oldFirst = this.first;
    if (this.size === 1) {
      this.first = null;
      this.last = null;
      this.size = 0;
    } else {
      this.first = this.first.next;
      this.size--;
    }
    oldFirst.next = null;
    return oldFirst;
  }
}

File: Python/5_DataStructures/4_Queues.py
class Simple_Queue(object):
	'''
	enqueue has Time Complexity of => O(1)
	dequeue has Time Complexity of => O(n) because of shift operator
	which moves all the values one index
	
	peek takes O(1)
	'''
	def __init__(self):
		self.queue = []

	def enqueue(self, val):
		self.queue.append(val)
		return True

	def dequeue(self):
		if len(self.queue) == 0: return False
		return self.queue.pop(0)

	def peek(self):
		if len(self.queue) == 0: return False
		return self.queue[0]



class Node(object): 
	def __init__(self, val):
		self.val = val
		self.next = None
		
		
class Queue(object):
	'''
	All the operations in this class has Time Complexity of O(1)
	'''
	def __init__(self):
		self.head = None
		self.tail = None
		self.length = 0

	def enqueue(self, val):
		new_node = Node(val)

		if not self.head:
			self.head = self.tail = new_node
		else:
			self.tail.next = new_node
			self.tail = new_node

		self.length += 1
		return True


	def dequeue(self):
		if self.length == 0: return False
		if self.length == 1:
			node = self.head
			self.head = self.tail = None
		else:
			node = self.head
			self.head = node

		self.length -= 1
		return node.val

	def peek(self):
		if self.length == 0: return False
		return self.head.val
		
		